File: transfers/day_to_day/simulator/driver.py
from dataclasses import dataclass

from common.progress import maybe_tqdm
from common.transactions import Transaction

from ..channels import build_channel_cdf
from ..dynamics import ContactsView, FavoritesView, PopulationDynamics
from ..environment import (
    FinancialObligations,
    Market,
    Parameters,
    TransactionEngine,
)
from .day import RunPlan, RunState, run_day
from .routing import RoutePolicy
from .spenders import prepare_spenders
from .sync import build_sensitivities, sync_dynamics
from .targets import target_txns


@dataclass(slots=True)
class Simulator:
    market: Market
    engine: TransactionEngine
    obligations: FinancialObligations
    params: Parameters

    def build_plan(self) -> RunPlan:
        market = self.market
        obligations = self.obligations
        params = self.params

        prepared_spenders = prepare_spenders(market, obligations)
        active_spenders = len(prepared_spenders)

        base_txns = (
            obligations.base_txns
            if obligations.base_txns_sorted
            else sorted(obligations.base_txns, key=lambda txn: txn.timestamp)
        )

        total_people = len(market.population.persons)
        per_person_daily_limit = int(market.events.max_per_person_per_day)
        person_limit = None if per_person_daily_limit <= 0 else per_person_daily_limit

        batch = (
            PopulationDynamics.from_persons(market.person_states)
            if total_people >= 500
            else None
        )

        favorites = FavoritesView(
            indices=market.merchants.fav_merchants_idx,
            counts=market.merchants.fav_k,
            cdf=market.merchants.merch_cdf,
            total=len(market.merchants.merchants.ids),
        )
        contacts = ContactsView(
            matrix=market.social.contacts,
            degree=market.social.degree,
            n_people=total_people,
        )

        return RunPlan(
            prepared_spenders=prepared_spenders,
            target_total_txns=target_txns(market, active_spenders),
            total_person_days=active_spenders * market.days,
            base_txns=base_txns,
            sensitivities=build_sensitivities(market),
            batch=batch,
            favorites=favorites,
            contacts=contacts,
            route_policy=RoutePolicy(
                prefer_billers_p=float(market.events.prefer_billers_p),
                max_retries=int(params.picking_rules.max_retries),
            ),
            person_limit=person_limit,
            channel_cdf=build_channel_cdf(market.events, market.merchants_cfg),
        )

    def run(self) -> list[Transaction]:
        if self.market.days <= 0:
            return []

        plan = self.build_plan()
        state = RunState(
            days_since_payday=[365] * len(self.market.population.persons),
        )

        for day_index in maybe_tqdm(
            range(self.market.days),
            desc="day-to-day",
            unit="day",
            leave=False,
        ):
            run_day(self, plan, state, day_index)

        if plan.batch is not None:
            sync_dynamics(plan.batch, self.market.person_states)

        return state.txns

File: src/spending/simulator/driver.cpp
#include "phantomledger/spending/simulator/driver.hpp"

#include "phantomledger/spending/routing/channel.hpp"
#include "phantomledger/spending/simulator/day.hpp"
#include "phantomledger/spending/spenders/prepare.hpp"
#include "phantomledger/spending/spenders/targets.hpp"
#include "phantomledger/transactions/record.hpp"

#include <algorithm>
#include <cstddef>

namespace PhantomLedger::spending::simulator {

namespace {

constexpr double kTxnReserveSlack = 1.05;

void sortChronological(std::vector<transactions::Transaction> &txns) {
  // Sort by (timestamp, source, target, amount) — the canonical key.
  std::sort(
      txns.begin(), txns.end(),
      transactions::Comparator{transactions::Comparator::Scope::fundsTransfer});
}

} // namespace

Simulator::Simulator(const market::Market &market, Engine &engine,
                     const obligations::Snapshot &obligations,
                     const dynamics::Config &dynamicsCfg,
                     const config::BurstBehavior &burst,
                     const config::ExplorationHabits &exploration,
                     const config::LiquidityConstraints &liquidity,
                     const config::MerchantPickRules &picking,
                     const math::seasonal::Config &seasonal,
                     double txnsPerMonth, double channelMerchantP,
                     double channelBillsP, double channelP2pP,
                     double unknownOutflowP, double baseExploreP,
                     double dayShockShape, double preferBillersP,
                     std::uint32_t personDailyLimit)
    : market_(market), engine_(engine), obligations_(obligations),
      dynamicsCfg_(dynamicsCfg), burst_(burst), exploration_(exploration),
      liquidity_(liquidity), picking_(picking), seasonal_(seasonal),
      txnsPerMonth_(txnsPerMonth), channelMerchantP_(channelMerchantP),
      channelBillsP_(channelBillsP), channelP2pP_(channelP2pP),
      unknownOutflowP_(unknownOutflowP), baseExploreP_(baseExploreP),
      dayShockShape_(dayShockShape), preferBillersP_(preferBillersP),
      personDailyLimit_(personDailyLimit),
      cohort_(market.population().count()) {
  const auto n = market.population().count();

  // Per-person paycheck sensitivity: drawn here from each persona.
  // `entity::behavior::Persona` nests this under `payday.sensitivity`;
  // the previous flat `paycheckSensitivity` accessor doesn't exist.
  sensitivities_.resize(n);
  for (std::uint32_t i = 0; i < n; ++i) {
    const auto person = static_cast<entity::PersonId>(i + 1);
    sensitivities_[i] = market.population().object(person).payday.sensitivity;
  }

  paydayPersonIndicesScratch_.reserve(n);
  momentumMultBuffer_.resize(n);
  dormancyMultBuffer_.resize(n);
  paycheckMultBuffer_.resize(n);
  dailyMultBuffer_.resize(n);
}

RunPlan Simulator::buildPlan() const {
  RunPlan plan{};

  plan.preparedSpenders = spenders::prepareSpenders(market_, obligations_);
  const std::uint32_t activeSpenders =
      static_cast<std::uint32_t>(plan.preparedSpenders.size());

  plan.targetTotalTxns = spenders::totalTargetTxns(
      txnsPerMonth_, activeSpenders, market_.bounds().days);

  plan.totalPersonDays =
      static_cast<std::uint64_t>(activeSpenders) * market_.bounds().days;

  plan.baseTxns = obligations_.baseTxns;
  plan.sensitivities = std::span<const double>(sensitivities_);

  if (personDailyLimit_ > 0) {
    plan.personLimit = personDailyLimit_;
  }

  plan.channelCdf = routing::buildChannelCdf(channelMerchantP_, channelBillsP_,
                                             channelP2pP_, unknownOutflowP_);

  plan.routePolicy = routing::Policy{
      .preferBillersP = preferBillersP_,
      .maxRetries = picking_.maxRetries,
  };

  plan.baseExploreP = baseExploreP_;
  plan.dayShockShape = dayShockShape_;

  return plan;
}

std::vector<transactions::Transaction> Simulator::run() {
  if (market_.bounds().days == 0) {
    return {};
  }

  RunPlan plan = buildPlan();

  // Reserve up front: target * 1.05 covers Poisson overshoot.
  const std::size_t reserveCapacity =
      static_cast<std::size_t>(plan.targetTotalTxns * kTxnReserveSlack);

  RunState state(market_.population().count(), reserveCapacity);

  for (std::uint32_t dayIndex = 0; dayIndex < market_.bounds().days;
       ++dayIndex) {
    runDay(market_, engine_, plan, state, cohort_, dynamicsCfg_, burst_,
           exploration_, liquidity_, seasonal_, dayIndex);
  }

  auto txns = std::move(state.txns());
  sortChronological(txns);
  return txns;
}

} // namespace PhantomLedger::spending::simulator
